Typescript definitions fail downstream when used

<!--
  ⚡️ katchow! We 💛 issues.

  Please - do not - remove this template.
  Please - do not - skip or remove parts of this template.
  Or your issue may be closed.

  👉🏽 Need help or support? Open a SUPPORT issue instead.
-->

- webpack-plugin-serve Version: `master`

### How Do We Reproduce?

:boom:  **Working on PR** 


### Expected Behavior
Typescript users should be able to `npm add -D webpack-plugin-serve`, and use it in a TS file without the need to do anything else. 


### Actual Behavior

This is a sample of the errors as-is once a usage test is added (reproduced from my config):

```
~/p/webpack-plugin-serve ❯❯❯ npm run test:typescript                                                                                                                                                                           ✘ 127 

> webpack-plugin-serve@0.6.0 test:typescript /Users/kross/projects/webpack-plugin-serve
> tslint --project tsconfig.json && tsc

index.d.ts:5:8 - error TS7016: Could not find a declaration file for module 'http-proxy-middleware'. '/Users/kross/projects/webpack-plugin-serve/node_modules/http-proxy-middleware/index.js' implicitly has an 'any' type.
  Try `npm install @types/http-proxy-middleware` if it exists or add a new declaration (.d.ts) file containing `declare module 'http-proxy-middleware';`

5 } from 'http-proxy-middleware';
         ~~~~~~~~~~~~~~~~~~~~~~~

index.d.ts:6:22 - error TS7016: Could not find a declaration file for module 'koa'. '/Users/kross/projects/webpack-plugin-serve/node_modules/koa/lib/application.js' implicitly has an 'any' type.
  Try `npm install @types/koa` if it exists or add a new declaration (.d.ts) file containing `declare module 'koa';`

6 import * as Koa from 'koa';
                       ~~~~~

index.d.ts:28:19 - error TS7006: Parameter 'res' implicitly has an 'any' type.

28     setHeaders?: (res, path, stats) => any;
                     ~~~

index.d.ts:28:24 - error TS7006: Parameter 'path' implicitly has an 'any' type.

28     setHeaders?: (res, path, stats) => any;
                          ~~~~

index.d.ts:28:30 - error TS7006: Parameter 'stats' implicitly has an 'any' type.

28     setHeaders?: (res, path, stats) => any;
                                ~~~~~


Found 5 errors.
```

--- a/.prettierrc
+++ b/.prettierrc
@@ -1,0 +1,5 @@
+printWidth: 100
+semi: true
+singleQuote: true
+trailingComma: all
+bracketSpacing: true

--- a/index.d.ts
+++ b/index.d.ts
@@ -1,99 +1,99 @@
-import { Url } from 'url';
+import { Url } from "url";
 import {
   Config as HttpProxyMiddlewareConfig,
   Proxy
-} from 'http-proxy-middleware';
-import * as Koa from 'koa';
+} from "http-proxy-middleware";
+import * as Koa from "koa";
 import {
   ServerOptions as Http2ServerOptions,
   SecureServerOptions as Http2SecureServerOptions
-} from 'http2';
-import { ServerOptions as HttpsServerOptions} from 'https';
-import { ZlibOptions } from 'zlib';
-import { Compiler } from 'webpack';
+} from "http2";
+import { ServerOptions as HttpsServerOptions } from "https";
+import { ZlibOptions } from "zlib";
+import { Compiler } from "webpack";
 
-declare module 'webpack-plugin-serve' {
-  interface CompressOptions extends ZlibOptions {
-    filter?: (content_type: string) => boolean;
-    threshold?: number;
-  }
+interface CompressOptions extends ZlibOptions {
+  filter?: (content_type: string) => boolean;
+  threshold?: number;
+}
 
-  interface KoaStaticOptions {
-    maxage?: number;
-    hidden?: boolean;
-    index?: string;
-    defer?: boolean;
-    gzip?: boolean;
-    br?: boolean;
-    setHeaders?: (res, path, stats) => any;
-    extensions?: Array<string> | boolean;
-  }
+interface KoaStaticOptions {
+  maxage?: number;
+  hidden?: boolean;
+  index?: string;
+  defer?: boolean;
+  gzip?: boolean;
+  br?: boolean;
+  setHeaders?: (res: any, path: any, stats: any) => any;
+  extensions?: Array<string> | boolean;
+}
 
-  type RewriteTo = (context: Context) => string;
+type RewriteTo = (context: Context) => string;
 
-  interface Context {
-    match: RegExpMatchArray;
-    parsedUrl: Url;
-  }
+interface Context {
+  match: RegExpMatchArray;
+  parsedUrl: Url;
+}
 
-  interface Rewrite {
-    from: RegExp;
-    to: string | RegExp | RewriteTo;
-  }
+interface Rewrite {
+  from: RegExp;
+  to: string | RegExp | RewriteTo;
+}
 
-  interface HistoryApiFallbackOptions {
-    disableDotRule?: true;
-    htmlAcceptHeaders?: string[];
-    index?: string;
-    logger?: typeof console.log;
-    rewrites?: Rewrite[];
-    verbose?: boolean;
-  }
+interface HistoryApiFallbackOptions {
+  disableDotRule?: true;
+  htmlAcceptHeaders?: string[];
+  index?: string;
+  logger?: typeof console.log;
+  rewrites?: Rewrite[];
+  verbose?: boolean;
+}
 
-  interface Builtins {
-    proxy: (args: HttpProxyMiddlewareConfig) => Proxy;
-    compress: (opts: CompressOptions) => void;
-    static: (paths: Array<String>, opts?: KoaStaticOptions) => void;
-    historyFallback: (opts: HistoryApiFallbackOptions) => void;
-    websocket: () => void;
-    four0four: (fn?: (ctx: Koa.Context) => void) => void;
-  }
+interface Builtins {
+  proxy: (args: HttpProxyMiddlewareConfig) => Proxy;
+  compress: (opts: CompressOptions) => void;
+  static: (paths: Array<String>, opts?: KoaStaticOptions) => void;
+  historyFallback: (opts: HistoryApiFallbackOptions) => void;
+  websocket: () => void;
+  four0four: (fn?: (ctx: Koa.Context) => void) => void;
+}
 
-  export interface WebpackPluginServeOptions {
-    client?: {
-      address?: string;
-      retry?: boolean;
-      silent?: boolean;
-    };
-    compress?: boolean;
-    historyFallback?: boolean | HistoryApiFallbackOptions;
-    hmr?: boolean;
-    host?: string | Promise<string>;
-    http2?: boolean | Http2ServerOptions | Http2SecureServerOptions;
-    https?: HttpsServerOptions;
-    liveReload?: boolean;
-    log?: {
-      level: string;
-      timestamp?: boolean;
-    };
-    middleware?: (app: Koa, builtins: Builtins) => void;
-    open?: boolean | {
-      wait?: boolean;
-      app?: string | ReadonlyArray<string>;
-    };
-    port?: number | Promise<number>;
-    progress?: boolean | 'minimal';
-    static?: string | Array<string>;
-    status?: boolean;
-    waitForBuild?: boolean;
-  }
+export interface WebpackPluginServeOptions {
+  client?: {
+    address?: string;
+    retry?: boolean;
+    silent?: boolean;
+  };
+  compress?: boolean;
+  historyFallback?: boolean | HistoryApiFallbackOptions;
+  hmr?: boolean;
+  host?: string | Promise<string>;
+  http2?: boolean | Http2ServerOptions | Http2SecureServerOptions;
+  https?: HttpsServerOptions;
+  liveReload?: boolean;
+  log?: {
+    level: string;
+    timestamp?: boolean;
+  };
+  middleware?: (app: Koa, builtins: Builtins) => void;
+  open?:
+    | boolean
+    | {
+        wait?: boolean;
+        app?: string | ReadonlyArray<string>;
+      };
+  port?: number | Promise<number>;
+  progress?: boolean | "minimal";
+  static?: string | Array<string>;
+  status?: boolean;
+  waitForBuild?: boolean;
+}
 
-  export class WebpackPluginServe {
-    constructor(opts?: WebpackPluginServeOptions);
-    attach(): {
-      apply(compiler: Compiler): void;
-    };
-    hook(compiler: Compiler): void;
+export class WebpackPluginServe {
+  constructor(opts?: WebpackPluginServeOptions);
+  attach(): {
     apply(compiler: Compiler): void;
-  }
+  };
+  hook(compiler: Compiler): void;
+  apply(compiler: Compiler): void;
 }

--- a/index.d.ts
+++ b/index.d.ts
@@ -1,16 +1,13 @@
-import { Url } from "url";
-import {
-  Config as HttpProxyMiddlewareConfig,
-  Proxy
-} from "http-proxy-middleware";
-import * as Koa from "koa";
+import { Url } from 'url';
+import { Config as HttpProxyMiddlewareConfig, Proxy } from 'http-proxy-middleware';
+import * as Koa from 'koa';
 import {
   ServerOptions as Http2ServerOptions,
-  SecureServerOptions as Http2SecureServerOptions
-} from "http2";
-import { ServerOptions as HttpsServerOptions } from "https";
-import { ZlibOptions } from "zlib";
-import { Compiler } from "webpack";
+  SecureServerOptions as Http2SecureServerOptions,
+} from 'http2';
+import { ServerOptions as HttpsServerOptions } from 'https';
+import { ZlibOptions } from 'zlib';
+import { Compiler } from 'webpack';
 
 interface CompressOptions extends ZlibOptions {
   filter?: (content_type: string) => boolean;
@@ -83,7 +80,7 @@ export interface WebpackPluginServeOptions {
         app?: string | ReadonlyArray<string>;
       };
   port?: number | Promise<number>;
-  progress?: boolean | "minimal";
+  progress?: boolean | 'minimal';
   static?: string | Array<string>;
   status?: boolean;
   waitForBuild?: boolean;

--- a/package.json
+++ b/package.json
@@ -26,8 +26,10 @@
     "faq:toc": "echo 'temporarily disabled' # markdown-toc -i .github/FAQ.md",
     "lint": "eslint --fix --cache lib test",
     "lint-staged": "lint-staged",
+    "pretest": "npm run test:typescript",
     "security": "npm audit",
-    "test": "ava"
+    "test": "ava",
+    "test:typescript": "tslint --project tsconfig.json && tsc"
   },
   "files": [
     "client.js",
@@ -61,10 +63,12 @@
   "devDependencies": {
     "@commitlint/cli": "^7.1.2",
     "@commitlint/config-conventional": "^7.1.2",
+    "@types/webpack": "^4.4.24",
     "ava": "^1.0.1",
     "cpy": "^7.0.1",
     "del": "^3.0.0",
     "del-cli": "^1.1.0",
+    "dtslint": "^0.4.2",
     "eslint": "^5.6.1",
     "eslint-config-shellscape": "^2.0.2",
     "execa": "^1.0.0",
@@ -77,6 +81,8 @@
     "prettier": "^1.14.3",
     "puppeteer": "^1.11.0",
     "standard-version": "^4.4.0",
+    "tslint": "^5.12.1",
+    "typescript": "^3.3.3",
     "webpack": "^4.28.1",
     "webpack-nano": "^0.6.1"
   },

--- a/package.json
+++ b/package.json
@@ -44,6 +44,7 @@
   "dependencies": {
     "@types/http-proxy-middleware": "^0.19.2",
     "@types/koa": "^2.0.48",
+    "@types/webpack": "^4.4.24",
     "chalk": "^2.4.1",
     "connect-history-api-fallback": "^1.5.0",
     "http-proxy-middleware": "^0.19.0",
@@ -65,7 +66,6 @@
   "devDependencies": {
     "@commitlint/cli": "^7.1.2",
     "@commitlint/config-conventional": "^7.1.2",
-    "@types/webpack": "^4.4.24",
     "ava": "^1.0.1",
     "cpy": "^7.0.1",
     "del": "^3.0.0",

--- a/package.json
+++ b/package.json
@@ -42,6 +42,8 @@
     "webpack": "^4.20.2"
   },
   "dependencies": {
+    "@types/http-proxy-middleware": "^0.19.2",
+    "@types/koa": "^2.0.48",
     "chalk": "^2.4.1",
     "connect-history-api-fallback": "^1.5.0",
     "http-proxy-middleware": "^0.19.0",

--- a/test/typescript/usage.test.ts
+++ b/test/typescript/usage.test.ts
@@ -23,7 +23,7 @@ function customUsage(wc: Configuration) {
           ctx.set("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
           ctx.set("Access-Control-Allow-Origin", "*");
         }),
-      static: "/", // needs to be the same as the output dir for the html-webpack-plugin
+      static: "/", // needs to be thet same as the output dir for the html-webpack-plugin
 
       //
       status: true,

--- a/test/typescript/usage.test.ts
+++ b/test/typescript/usage.test.ts
@@ -1,0 +1,38 @@
+import { Configuration } from "webpack";
+import { WebpackPluginServe } from "webpack-plugin-serve";
+
+function customUsage(wc: Configuration) {
+  // ← important: this is required, where the magic happens in the browser
+  (wc.entry as string[]).push("webpack-plugin-serve/client");
+
+  // ← important: webpack and the server will continue to run in watch mode
+  wc.watch = true;
+
+  wc.plugins!.push(
+    new WebpackPluginServe({
+      compress: true,
+      historyFallback: true,
+      host: "0.0.0.0",
+      port: 3808,
+      liveReload: true,
+      // set CORS headers
+      middleware: (app: any, builtins: any) =>
+        app.use(async (ctx: any, next: any) => {
+          await next();
+          ctx.set("Access-Control-Allow-Headers", "*");
+          ctx.set("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
+          ctx.set("Access-Control-Allow-Origin", "*");
+        }),
+      static: "/", // needs to be the same as the output dir for the html-webpack-plugin
+
+      //
+      status: true,
+      progress: true
+    })
+  );
+
+  // override the publicPath since production may be set to assets host
+  wc.output!.publicPath = "/";
+
+  return wc;
+}

--- a/test/typescript/usage.test.ts
+++ b/test/typescript/usage.test.ts
@@ -1,9 +1,9 @@
-import { Configuration } from "webpack";
-import { WebpackPluginServe } from "webpack-plugin-serve";
+import { Configuration } from 'webpack';
+import { WebpackPluginServe } from 'webpack-plugin-serve';
 
 function customUsage(wc: Configuration) {
   // ← important: this is required, where the magic happens in the browser
-  (wc.entry as string[]).push("webpack-plugin-serve/client");
+  (wc.entry as string[]).push('webpack-plugin-serve/client');
 
   // ← important: webpack and the server will continue to run in watch mode
   wc.watch = true;
@@ -12,27 +12,27 @@ function customUsage(wc: Configuration) {
     new WebpackPluginServe({
       compress: true,
       historyFallback: true,
-      host: "0.0.0.0",
+      host: '0.0.0.0',
       port: 3808,
       liveReload: true,
       // set CORS headers
       middleware: (app: any, builtins: any) =>
         app.use(async (ctx: any, next: any) => {
           await next();
-          ctx.set("Access-Control-Allow-Headers", "*");
-          ctx.set("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
-          ctx.set("Access-Control-Allow-Origin", "*");
+          ctx.set('Access-Control-Allow-Headers', '*');
+          ctx.set('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
+          ctx.set('Access-Control-Allow-Origin', '*');
         }),
-      static: "/", // needs to be thet same as the output dir for the html-webpack-plugin
+      static: '/', // needs to be thet same as the output dir for the html-webpack-plugin
 
       //
       status: true,
-      progress: true
-    })
+      progress: true,
+    }),
   );
 
   // override the publicPath since production may be set to assets host
-  wc.output!.publicPath = "/";
+  wc.output!.publicPath = '/';
 
   return wc;
 }

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,0 +1,14 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es5",
+    "lib": ["es6"],
+    "moduleResolution": "node",
+    "forceConsistentCasingInFileNames": true,
+    "strict": true,
+    "noEmit": true,
+    "baseUrl": ".",
+    "paths": { "webpack-plugin-serve": ["./index.d.ts"] }
+  },
+  "include": ["./index.d.ts", "./test/**/*.ts*"]
+}

--- a/tslint.json
+++ b/tslint.json
@@ -1,0 +1,7 @@
+{
+  "defaultSeverity": "error",
+  "extends": "dtslint/dtslint.json",
+  "rules": {
+    "semicolon": true
+  }
+}

